src/crimson/osd/pg_recovery.cc

```diff
@@ -0,0 +1,466 @@
+    auto pm = pg->peering_state.get_peer_missing().find(peer);
+    assert(pm != pg->peering_state.get_peer_missing().end());
+
+    size_t m_sz = pm->second.num_missing();
+
+    crimson::get_logger(ceph_subsys_osd).debug(
+	"{}: peer osd.{} missing {} objects",
+	__func__,
+	peer,
+	m_sz);
+    crimson::get_logger(ceph_subsys_osd).trace(
+	"{}: peer osd.{} missing {}", __func__,
+	peer, pm->second.get_items());
+
+    // recover oldest first
+    const pg_missing_t &m(pm->second);
+    for (auto p = m.get_rmissing().begin();
+	 p != m.get_rmissing().end() && started < max_to_start;
+	 ++p) {
+      const auto &soid = p->second;
+
+      if (pg->peering_state.get_missing_loc().is_unfound(soid)) {
+	crimson::get_logger(ceph_subsys_osd).debug(
+	    "{}: object {} still unfound", __func__, soid);
+	continue;
+      }
+
+      const pg_info_t &pi = pg->peering_state.get_peer_info(peer);
+      if (soid > pi.last_backfill) {
+	if (!pg->recovery_backend->is_recovering(soid)) {
+	  crimson::get_logger(ceph_subsys_osd).error(
+	    "{}: object {} in missing set for backfill (last_backfill {})"
+	    " but not in recovering",
+	    __func__,
+	    soid,
+	    pi.last_backfill);
+	  ceph_abort();
+	}
+	continue;
+      }
+
+      if (pg->recovery_backend->is_recovering(soid)) {
+	crimson::get_logger(ceph_subsys_osd).debug(
+	    "{}: already recovering object {}", __func__, soid);
+	continue;
+      }
+
+      if (pg->peering_state.get_missing_loc().is_deleted(soid)) {
+	crimson::get_logger(ceph_subsys_osd).debug(
+	    "{}: soid {} is a delete, removing", __func__, soid);
+	map<hobject_t,pg_missing_item>::const_iterator r =
+	  m.get_items().find(soid);
+	started += pg->prep_object_replica_deletes(
+	  soid, r->second.need, out);
+	continue;
+      }
+
+      if (soid.is_snap() &&
+	  pg->peering_state.get_pg_log().get_missing().is_missing(
+	    soid.get_head())) {
+	crimson::get_logger(ceph_subsys_osd).debug(
+	    "{}: head {} still missing on primary",
+	    __func__, soid.get_head());
+	continue;
+      }
+
+      if (pg->peering_state.get_pg_log().get_missing().is_missing(soid)) {
+	crimson::get_logger(ceph_subsys_osd).debug(
+	    "{}: soid {} still missing on primary", __func__, soid);
+	continue;
+      }
+
+      crimson::get_logger(ceph_subsys_osd).debug(
+	"{}: recover_object_replicas({})",
+	__func__,
+	soid);
+      map<hobject_t,pg_missing_item>::const_iterator r = m.get_items().find(
+	soid);
+      started += pg->prep_object_replica_pushes(
+	soid, r->second.need, out);
+    }
+  }
+
+  return started;
+}
+
+template<typename T>
+size_t PGRecovery<T>::start_backfill_ops(
+  size_t max_to_start,
+  std::vector<crimson::osd::blocking_future<>> *out)
+{
+  T* pg = static_cast<T*>(this);
+  crimson::get_logger(ceph_subsys_osd).debug(
+    "{}({}): bft={} lbs={} {}",
+    __func__,
+    pg->peering_state.get_backfill_targets(),
+    last_backfill_started,
+    new_backfill ? "" : "new_backfill");
```

Comment:
where is `new_backfill` set?

Reply:
I guess it isn't set in `crimson` yet as this is a part of backfill. In classical `new_backfill = true` happens in `PrimaryLogPG::on_activate_complete()`.